core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -450,9 +450,9 @@ private[log] class Cleaner(val id: Int,
                              retainDeletes: Boolean,
                              maxLogMessageSize: Int,
                              stats: CleanerStats) {
-
-    val logCleanerFilter = new LogEntryFilter {
-      def shouldRetain(logEntry: LogEntry): Boolean = shouldRetainMessage(source, map, retainDeletes, logEntry, stats)
+    val logCleanerFilter = new RecordFilter {
+      def shouldRetain(record: Record): Boolean =
+        record.isControlRecord || shouldRetainMessage(source, map, retainDeletes, record, stats)
```

Comment:
Nit: it may make sense to keep all the logic in one place. `shouldRetainMessage` is only called from this method, so is there any reason not to just add the first check there as well?